### ISSUE TITLE
Duplicate table row

### ISSUE THREAD

**Issue by revolter** [sqlitebrowser/sqlitebrowser#577]
It would be a nice addition a duplicate row functionality.

Similar to the column name header right click popup (`Show rowid column`, `Edit display format`, e.g.), a popup could appear when right clicking the left column numbers which would present a `Duplicate row` menu item which, after clicking it, would insert another row with the same column values, except the autoincrement one.

**Comment by tonypdmtr** [comment]
A 'ditto' function!  I second!  And don't forget keyboard fans (CTRL-" or something, like in some spreadsheets).

Actually, it'd be nice to have two 'ditto' functions.  One to copy just the cell above, and one to copy the whole row.  (OK, maybe the whole row can be done with the mouse only as one can repeat the keyboard way for each cell of interest.)

**Comment by innermous** [comment]
@tonypdmtr are you sure Ctrl-" is common shortcut for that?
I picked up the task btw.

**Comment by tonypdmtr** [comment]
Yes, Excel uses it, for one (see official link below).  The actual combination is CTRL together with a single quote (no SHIFT is needed to get the double quote) but it's referred to as CTRL-" (double-quote).  This is supposedly easy to remember as it resembles the '' (two short vertical bars) used for a 'ditto' notation.  Have a look here: https://en.wikipedia.org/wiki/Ditto_mark and here: https://support.office.com/en-us/article/Excel-shortcut-and-function-keys-1798d9d5-842a-42b8-9c99-9b7213f0040f (look down the page for the CTRL-'  single quote, and its description)

**Comment by innermous** [comment]
@tonypdmtr what about UNIQUE fields? 
I am thinking about UX design of that feature. Should button be greyed out in those cases? In other case it will yield error with failed constraint.

**Comment by justinclift** [comment]
From a logical perspective, having a button near the add/delete record ones makes sense.  Is that how people needing the feature would use it though?  (I'm not sure)

If it's easy to do, maybe try first with that and we can get some people to test it out to see if it "works ok" in actual practise?

**Comment by tonypdmtr** [comment]
The unique fields issue is irrelevant in the sense that the ditto operation simply copies the cell (or row) above to this one.  This is either an allowed thing or not.

What I mean is the operation is no different from someone trying to actually type in the respective field values.  This will be either allowed, or not.  How it displays an error is no different in either case.

Regarding the UI issue, (I'm a keyboard person so it the CTRL-' covers me).  My idea about it, however, would be to have a right click on the row/field you want to copy and a popup window to give two options: 'Copy Row' and 'Copy Field'.  No special buttons.  What do you think?

**Comment by justinclift** [comment]
Yeah, the right click approach sounds useful without cluttering up the UI right away. :smile:

**Comment by tonypdmtr** [comment]
Just to add a couple more details of how I envision this:

1. CTRL-' (single quote) can be for cell/field copy, and CTRL-" (double quote) for row copy.  Easy to remember: One quote does less than two quotes.

2. Once the 'ditto' operation is complete, the cursor could automatically move to:

(a) the cell/field to the right, when a single cell/field is copied.  (If on last column, go to first column -- i.e., wrap around.)  This will allow repeated CTRL-' to very quickly copy multiple fields, even when not the whole row is needed.

-- or --

(b) the first cell/field, when a whole row is copied, to allow possible edits.

What do you think?

**Comment by innermous** [comment]
@tonypdmtr you can ditto field one time. After that it's ambiguous where to ditto another field, cuz new row is already created.
IMO only ditto row matters.

**Comment by tonypdmtr** [comment]
Ambiguous?  Ditto always works in relation to the field above.  So, wherever the cursor is, that's where (the column) you get your value from.

IMO, if you have to keep only one, then only ditto single field matters.  Because, it will never be the case that a new row is needed that has the exact same information for each and every column.  On the other hand, it's quite common to have the same values repeated for one or more columns in different rows.

**Comment by innermous** [comment]
@tonypdmtr ow...
I thought this option should duplicate current row in the end of table... Now I see I imagined everything wrong, sorry.

**Comment by justinclift** [comment]
Heh Heh Heh, that's why we discuss stuff. :smile:

**Comment by tonypdmtr** [comment]
@innermous You thought right except that it doesn't necessarily have to be the whole row.  If you also take into account the current column (where the cursor is, or mouse click happens), it's the same thing.  You just use both grid coordinates, not just the row, from where to get the source data.

Copying the whole row is convenient when someone wants to keep all but one or a few fields the same (happens less often).

Copying a single field is convenient when one wants to clone one or few fields only (happens more often).

Once the copy is done, the user will most certainly need to work on the fields that need to be different, so it'd be nice to stay on that new row and be allowed to continue editing as usual.

**Comment by innermous** [comment]
@tonypdmtr u don't understand what mean. Everything you say is right. But real thing is, what happens when we have K rows and we ditto row N? There are 3 possible outcomes:
1. Some data from N (column or whole row whaaatever it doesn't matter now) is copied to K + 1
2. Data from N copies to N + 1 (following item)
3. Data from previous item (N - 1) dittos to N.
You see how ambiguous the whole thing is? We must consider which one of these we need, then we can proceed further to details and implementation.

**Comment by tonypdmtr** [comment]
Let's make simple then.

You press 'New Record' in browser, and a new row appears ready to accept your data.
You position the cursor to some cell in that new row.
You do DITTO function.
The contents of the immediately above row for the same column is copied into the current cell as if typed in by hand.
Then, ideally, the cursor on the current (new record) row, moves to the next column.

Did I answer your question?

**Comment by innermous** [comment]
@tonypdmtr well, at least I see whole picture now.

**Comment by revolter** [comment]
I too got lost in this discussion. My initial feature request and actual use case scenario was to duplicate any table row, and, of course, every column value, except, unique fields.

**Comment by revolter** [comment]
My use case was quite different from what @tonypdmtr sees.

This:

> Copying the whole row is convenient when someone wants to keep all but one or a few fields the same (happens less often).

was false in my case. I did want to copy an entire row multiple times, and only change 1-2 fields.

**Comment by revolter** [comment]
Knowing the "A picture is worth a thousand words" idiom, here are some pictures to better explain myself:

Before:

![screen shot 2016-05-16 at 15 23 14](https://cloud.githubusercontent.com/assets/5748627/15289388/87c44be0-1b7a-11e6-816b-c3d624ef1cc1.png)

Right click in the red circle and select "Duplicate row":

![screen shot 2016-05-16 at 15 23 14_1](https://cloud.githubusercontent.com/assets/5748627/15289511/5245dd8e-1b7b-11e6-84ae-06225650fdf9.png)

After:

![screen shot 2016-05-16 at 15 31 43](https://cloud.githubusercontent.com/assets/5748627/15289516/5bafe6da-1b7b-11e6-840d-8c5fd31e0215.png)

**Comment by justinclift** [comment]
Just to confirm, the `id` column is in the pics is autogenerated yeah? (that's why it's value isn't copied?)

**Comment by revolter** [comment]
Correctly.

**Comment by tonypdmtr** [comment]
@revolter Hmm. How exactly is what I said different from what you just said?  You say the same thing only worded differently.  (Unless you're referring to the 'happens less often' part. But this is a matter of opinion.)

> My use case was quite different from what @tonypdmtr sees.
> 
> This:
> 
>     Copying the whole row is convenient when someone wants to keep all but one or a few fields the same (happens less often).
> 
> was false in my case. I did want to copy an entire row multiple times, and only change 1-2 fields.

**Comment by revolter** [comment]
@tonypdmtr, I may have got it wrong, sorry. Not a native english​ speaker.​

**Comment by innermous** [comment]
So ditto field always dittos current field into the very last record (created with "New record"), right? So it can be called for several fields.

**Comment by justinclift** [comment]
It sounds like it should be used for all fields in the current row, that don't have a default (or autoincrement) value.

**Comment by innermous** [comment]
@justinclift was speaking about Ctrl-' (ditto single field). Everything is clear with ditto row.

**Comment by justinclift** [comment]
Oops, sorry. :wink:

**Comment by innermous** [comment]
I have this working, not sure implementation is good though.

**Comment by tonypdmtr** [comment]
@innermous I agree about the CTRL-" ('ditto' for row) that it appears to be simpler when done directly (i.e., without having to do Add Record first).

As for copying the same row several times by repeated CTRL-" presses, I would think that in most cases, it would be practically impossible.  A correctly designed database would not expect the exact same row to be repeated without any changes in (at least) one field.  So, some editing (automatic or manual) has to happen for each new record before going to the next one.  If the change is automatic (like an auto-increment field) then that's not an issue.  But I would think this to be the exception rather than the rule for most tables.

**Comment by innermous** [comment]
@tonypdmtr okay, then leaving it as is.

**Comment by MKleusberg** [comment]
@revolter I've just merged @innermous's PR for this. Do you have the time to try tomorrow's nightly build and report if it's working for you? :smiley:

**Comment by revolter** [comment]
@MKleusberg Sure thing :D

**Comment by revolter** [comment]
@MKleusberg It's working great, thanks a lot @innermous!
